Added Italian language support for the LessMSI GUI app

mega5800 opened the pull request:
:+1::tada: First off, thanks for taking the time to contribute! :tada::+1:

Please fill out the following checklist:

- [x] Added tests for any bug fixes that changed existing code
- [x] Added tests for any new behavior
- [x] All unit tests are passing (please make sure all tests are passing for your branch/PR at https://ci.appveyor.com/project/activescott/lessmsi/history )


If you need any help at all, feel free to submit the PR and @-mention activescott and I'll be happy to assist!

Hello @activescott.

I took care of this [ticket](https://github.com/activescott/lessmsi/issues/203) and added Italian text support for the LessMSI GUI app.

Here are some screenshots of my work:
![image](https://github.com/user-attachments/assets/3439d41f-17cb-4325-ba26-4c81e6de363d)

![image](https://github.com/user-attachments/assets/b887581e-03cd-4112-9c59-9470e052b1a7)

@bovirus, thank you for your help with Italian translations.

Thank you.

mega5800 commented:
@activescott can you please help me with the raised Codacy error:
![image](https://github.com/user-attachments/assets/0d46eee3-6968-48ee-aaf7-d184e7a6579e)

The ```partial``` keywork is required here, as this is a Form class:
![image](https://github.com/user-attachments/assets/d6445cac-0445-46b2-afc2-82cf0f8e6d92)

mega5800 commented:
@activescott any updates?

Thank you.

bovirus commented:
@mega5800 

How can help you with the code?

mega5800 commented:
Thank you for your help suggestion @bovirus.

The issue is with the Codacy test rather than with the code itself.
I am waiting for response from Scott about this issue.

activescott commented:
@mega5800 I ignored the codacy issue. Can you also try removing "partial" in that one place and see if it compiles. I can't remember if C# allows it to be omitted in that case or not - but that is what Codacy/Sonar# is implying. I don't have my windows dev box handy right now, but I'll investigate this more closely here soon.

activescott commented on src/LessMsi.Gui/AboutBox.cs:
```diff
@@ -96,15 +99,48 @@ public string AssemblyCompany
-		}
+		}
 		#endregion
-
```
@mega5800 I'd rather we just not translate the rtf file. Lets just leave it English. This complexity is not worth it.

mega5800 replied:
In this pull request, you are able to see a brand new logic I created for improving how our GUI app handles RTF files across different locales.

Instead of relying on a hardcoded RTF file in English, I've developed a generalized RTF template system. This template contains locale-independent content, such as names and links, while using placeholders for phrases that vary based on the user's language. 
My logic dynamically fills these placeholders with the appropriate locale-specific phrases, as defined in our strings file.

This approach not only streamlines our localization process but also ensures that all users, regardless of their locale, experience a polished and consistent interface. I believe this enhancement will contribute to a more personalized and professional user experience.

activescott replied:
I saw the code. That is actually what concerns me. I'd rather not have that added logic/complexity for the project to maintain for the content there that I do not think is important. The goal of the translation is to make it usable in other languages, and the about text - especially the few strings that are translatable there isn't important enough to translate to justify that added complexity (for example, it needs tests to prevent regressions, translations need to concern with RTF-ecoded content, their is new file I/O logic that needs to consider permissions, and odd paths, etc. etc.).

mega5800 replied:
I understand your point and will revert this change.

activescott commented on src/LessMsi.Gui/MainForm.cs:
```diff
@@ -626,8 +637,11 @@ private void InitializeComponent()
             this.btnExtractStreamFiles.Size = new System.Drawing.Size(129, 23);
             this.btnExtractStreamFiles.TabIndex = 0;
-            this.btnExtractStreamFiles.Text = "Extract Stream Files...";
+            this.btnExtractStreamFiles.Text = $"{Strings.ExtractStreamFiles}...";
```
I think you should put the `...` in the translation itself here rather than combining translated text and a hardcoded string here. That may not be appropriate in all languages.

mega5800 commented:
@activescott I tried removing the ```partial``` keyword but got this error
![image](https://github.com/user-attachments/assets/33edc860-f919-4278-bb57-c1696ccd01d0)

Is there an option of ignoring this Codacy issue while merging this pull request to the main repo?
This pull request only introduces GUI changes, without logic changes, and this is only a minor Codacy issue.

Thank you.

activescott commented on src/LessMsi.Gui/MainForm.cs:
```diff
@@ -647,7 +661,7 @@ private void InitializeComponent()
             this.lblStream.Size = new System.Drawing.Size(44, 15);
             this.lblStream.TabIndex = 11;
-            this.lblStream.Text = "Stream";
+            this.lblStream.Text = $"{Strings.Stream}";
```
```suggestion
            this.lblStream.Text = Strings.Stream;

```

activescott commented on src/LessMsi.Gui/MainForm.cs:
```diff
@@ -722,13 +736,13 @@ private void InitializeComponent()
             this.fileToolStripMenuItem.Name = "fileToolStripMenuItem";
             this.fileToolStripMenuItem.Size = new System.Drawing.Size(37, 20);
-            this.fileToolStripMenuItem.Text = "&File";
+            this.fileToolStripMenuItem.Text = $"&{Strings.File}";
```
I think the `&` should be added to the translation itself here - maybe rename it to `FileMenuLabel` or something if it is confusing to have "File" there. The I doubt the first letter will make sense in all languages as the keyboard accelerator.

mega5800 replied:
added ```FileMenuLabel``` string as you requested

activescott commented on src/LessMsi.Gui/MainForm.cs:
```diff
@@ -722,13 +736,13 @@ private void InitializeComponent()
             this.openToolStripMenuItem.Name = "openToolStripMenuItem";
             this.openToolStripMenuItem.Size = new System.Drawing.Size(152, 22);
-            this.openToolStripMenuItem.Text = "&Open";
+            this.openToolStripMenuItem.Text = $"&{Strings.Open}";
```
Same as above - put the `&` character into the translations themsleves incase a different keyboard accelerator makes more sense in other languages.

activescott commented on src/LessMsi.Gui/MainForm.cs:
```diff
@@ -753,7 +767,7 @@ private void InitializeComponent()
             this.exitToolStripMenuItem.ShortcutKeys = ((System.Windows.Forms.Keys)((System.Windows.Forms.Keys.Alt | System.Windows.Forms.Keys.F4)));
             this.exitToolStripMenuItem.Size = new System.Drawing.Size(152, 22);
-            this.exitToolStripMenuItem.Text = "E&xit";
+            this.exitToolStripMenuItem.Text = $"{Strings.Exit}";
```
Is the `&` in the translation in this one or did we lose the `Alt+X` keyboard accelerator ?

mega5800 replied:
The ```&``` is in both the english and italian translations:
![image](https://github.com/user-attachments/assets/ba9c662b-01a4-46b0-8b8b-942a25401eb4)


![image](https://github.com/user-attachments/assets/4fa5a551-baf5-4ae5-97b7-d15ec4f7ff8d)

activescott commented on src/LessMsi.Gui/MainForm.cs:
```diff
@@ -765,14 +779,14 @@ private void InitializeComponent()
             this.editToolStripMenuItem.Name = "editToolStripMenuItem";
             this.editToolStripMenuItem.Size = new System.Drawing.Size(39, 20);
-            this.editToolStripMenuItem.Text = "&Edit";
+            this.editToolStripMenuItem.Text = $"&{Strings.Edit}";
```
same as above - add `&` to the translation

activescott commented on src/LessMsi.Gui/MainForm.cs:
```diff
@@ -765,14 +779,14 @@ private void InitializeComponent()
             this.copyToolStripMenuItem.ShortcutKeys = ((System.Windows.Forms.Keys)((System.Windows.Forms.Keys.Control | System.Windows.Forms.Keys.C)));
             this.copyToolStripMenuItem.Size = new System.Drawing.Size(170, 22);
-            this.copyToolStripMenuItem.Text = "&Copy";
+            this.copyToolStripMenuItem.Text = $"&{Strings.Copy}";
```
same as above - add `&` to the translation

activescott commented on src/LessMsi.Gui/MainForm.cs:
```diff
@@ -793,14 +807,14 @@ private void InitializeComponent()
             this.searchFileToolStripMenuItem.ShortcutKeys = ((System.Windows.Forms.Keys)((System.Windows.Forms.Keys.Control | System.Windows.Forms.Keys.F)));
             this.searchFileToolStripMenuItem.Size = new System.Drawing.Size(170, 22);
-            this.searchFileToolStripMenuItem.Text = "Search File";
+            this.searchFileToolStripMenuItem.Text = $"{Strings.SearchFile}";
```
```suggestion
            this.searchFileToolStripMenuItem.Text = Strings.SearchFile;

```

activescott commented on src/LessMsi.Gui/MainForm.cs:
```diff
@@ -793,14 +807,14 @@ private void InitializeComponent()
             this.aboutToolStripMenuItem.Name = "aboutToolStripMenuItem";
             this.aboutToolStripMenuItem.Size = new System.Drawing.Size(52, 20);
-            this.aboutToolStripMenuItem.Text = "About";
+            this.aboutToolStripMenuItem.Text = $"{Strings.About}";
```
```suggestion
            this.aboutToolStripMenuItem.Text = Strings.About;

```

activescott commented on src/LessMsi.Gui/MainForm.cs:
```diff
@@ -914,7 +928,7 @@ private void btnExtract_Click(object sender, EventArgs e)
 			if (fileGrid.SelectedRows.Count == 0)
 			{
-				ShowUserMessageBox("Please select some or all of the files to extract them.");
+				ShowUserMessageBox($"{Strings.SelectFilesMessage}.");
```
```suggestion
				ShowUserMessageBox(Strings.SelectFilesMessage);

```
Punctuation (in this case the period) should be in the translation itself.

activescott commented on src/LessMsi.Gui/MainFormPresenter.cs:
```diff
@@ -131,16 +132,16 @@ private void ViewFiles(Database msidb)
 						ViewLeakedAbstraction.fileGrid.DataSource = fileDataSource;
 						View.AutoSizeFileGridColumns();
-						Status(fileDataSource.Count + " files found.");
+						Status(fileDataSource.Count + $" {Strings.FilesFoundStatus}.");
```
add punctuation to the translation itself

activescott commented on src/LessMsi.Gui/MainFormPresenter.cs:
```diff
@@ -330,7 +331,7 @@ public void LoadTables()
 					try
 					{
-						Status("Loading list of tables...");
+						Status($"{Strings.LoadingTablesStatus}...");
```
add punctutation to the translation itself (`...` in this case)

activescott commented on src/LessMsi.Gui/MainFormPresenter.cs:
```diff
@@ -378,7 +379,7 @@ public void OnSelectedStreamChanged()
 					if (foundStream == null)
 					{
-						View.ShowUserError("Could not find stream for CAB '{0}'", View.SelectedStreamInfo.Name);
+						View.ShowUserError($"{Strings.FindStreamError} CAB '{0}'", View.SelectedStreamInfo.Name);
```
put the placeholder (`{0}`) in to the translation string - this was done appropriately in other places.

activescott commented on src/LessMsi.Gui/MainFormPresenter.cs:
```diff
@@ -433,7 +434,7 @@ private void UpdateMSiTableGrid(Database msidb, string tableName)
 				return;
 
-			Status(string.Concat("Processing Table \'", tableName, "\'."));
+			Status(string.Concat($"{Strings.ProcessingTableStatus} \'", tableName, "\'."));
```
minor, but ideally convert this to a `string.format` expression and put the placeholder into the string so that the translator can decide how to "concat" the text appropriately.

activescott commented on src/LessMsi.Gui/Resources/Languages/Strings.it.Designer.cs:
is this supposed to be here?

mega5800 replied:
I just tested this matter and deleted this file.
I was able to compile the GUI app and run it without issues.

activescott commented on src/LessMsi.Gui/Windows.Forms/SearchPanel.Designer.cs:
```diff
@@ -56,7 +58,7 @@ private void InitializeComponent() {
 			this.label1.Size = new System.Drawing.Size(46, 18);
 			this.label1.TabIndex = 2;
-			this.label1.Text = "Search:";
+			this.label1.Text = $"{Strings.Search}:";
```
Add the punctuation to the translation

activescott commented on src/LessMsi.Gui/Resources/Languages/Strings.Designer.cs:
How is this picked up by the msbuild script? Can you also please add a unit test that loads .it language and confirm that some text was translated so that is always tested in CI. You need not test every string, maybe just a couple important ones.

mega5800 replied:
I am able to switch the locale of the application by running the next line of code:
```Thread.CurrentThread.CurrentUICulture = new CultureInfo("it");```

I believe this is something the environment takes into account when running the app.
I agree about testing the translations.
Is there an example of tests for the GUI app?

Thank you.

mega5800 replied:
Found a way of adding GUI tests to the existing tests

activescott reviewed (changes requested):
Looks pretty good, but a couple minor changes need made and a test for CI to make sure that the builds are always building properly translated apps. 

A quick though on tests: Maybe after the build is done, some of the CLI could be run to confirm it returned translated text? Just a an idea.

mega5800 commented:
@activescott thank you for your comments.
I took care of most of them and left a few comments of my own.
I'd love to get some extra feedback from you.

Thank you.

mega5800 commented:
Hello @activescott.

Following your advice, I added some tests for GUI in different languages (English and Italian).
These tests run successfully on my machine, but once I push them to the app veyor server the Italian test fail.
I believe my locale changing logic isn’t working properly on the app veyor server.
What can we do in such case?

Thank you.

mega5800 commented:
@activescott any updates?

Thank you.

activescott commented:
I'll look at the test case. I see the code but I'll do some debugging myself too

activescott commented on src/LessMsi.Gui/MainForm.cs:
```diff
@@ -237,23 +240,23 @@ public void AddPropertyGridColumn(string boundPropertyName, string headerText)
 		private Label label2;
 		private Panel panel1;
-		private Button btnExtract;
+		public Button btnExtract;
```
why did several of these buttons become public? These should almost certainly be private.

activescott replied:
Aaahhh. I see it was for the test now.

bovirus commented:
@activescott 

Could be possible to release a temporary build to test translation?

mega5800 commented:
@activescott, as you requested I reverted the changes I made to the AboutBox logic and updated the pull request.

I believe the only thing left is to sort the issue with ```LessMsi.Tests.GUITests.CheckUIStrings```.
@bovirus, I believe that once all the tests are executed successfully, Scott can publish a new version.

If you will find any issues with the translations at that point, I will gladly assist you.

mega5800 commented:
@activescott any updates?

Thank you.

mega5800 commented:
Hello @activescott.

I'd be grateful if you could check the testing I’ve been having.
This is the last bit in this pull request, and I’d like to start working on other tickets.

Thank you.

activescott commented:
I am looking. Sorry for the delay

activescott commented:
If you run it locally using the build.bat, test.bat scripts locally it also fails. Those scripts are really the only reliable way to build and test the project as Visual Studio has various "magic" that isn't always obvious. In this case the magic is that it is running commans to build the resources, and those aren't included in the build and the deployed files. I'll take a stab at adding that and push it.

mega5800 commented:
@activescott, thank you for examining this issue. 

I would like to mention that the only test that currently fails is the locale test, where I am changing the locale and testing the texts displayed in different buttons across the GUI app.
As you can see from the screenshot below, I am failing the Italian checks:
![image](https://github.com/user-attachments/assets/4cbe76b5-25dc-4f01-9398-b87f8bff2db1)

Perhaps, I can't change the locale on the AppVeyor server.

bovirus commented:
@mega5800

Italian should be "Seleziona &tutto"

mega5800 commented:
@bovirus 
I matched the format in which the original English text is shown, where each word is in capital case.

bovirus commented:
@mega5800 

In Italian the grammtical rule is different.
Only first letter of a paragraph (except own name) could be in uppercase.

And "is "Seleziona tutto" and not "Selziona tutto".

activescott commented:
I just pushed 62138c134581017532fed58b47e53e41e908f712 which fixed it locally when runnig build.bat and test.bat. Lets just double-check it passes in CI...

activescott commented:
Didn't work. I'll leave some notes here in case I don't figure this out here shortly before I have to take off for a bit:

https://ci.appveyor.com/project/activescott/lessmsi/builds/50698208#L250 shows that it did attempt to generate the it resources sattelite assembly.

https://ci.appveyor.com/project/activescott/lessmsi/builds/50698208#L266 shows it was generated and copied to the `LessMsi.Gui\bin\Release\it\lessmsi-gui.resources.dll` folder.

VS also copied it into the tests directory: https://ci.appveyor.com/project/activescott/lessmsi/builds/50698208#L311

The msbuild script did NOT copy it to the deploy directory which I would expect to have seen two lines after https://ci.appveyor.com/project/activescott/lessmsi/builds/50698208#L437

🤔

activescott commented:
Well the good news (?) is that it fails locally now for me :) it succeeded earlier though so apparently I had some msbuild syntax that worked and broke it lol

activescott commented:
i just tried another simpler approach in msbuild. Hope this works! 🤞

activescott commented:
looks like the it directory isn't in the zip though. So couple things left here that I know of:

- [ ] ensure the \it\*.resources files are included inthe zip (there is only one for now)
- [ ] ensure that the nupkg install also extracts the it subdir and the resources file in it

mega5800 commented:
@activescott, thank you for your help.
These points you mentioned, do you want me to take care of them?

If so, I'd be grateful if you could provide some information, since I am not familiar with their content.

@bovirus, I believe we are getting closer to releasing  a version with the Italian captions.
I'd like to remind that all the Italian captions were provided by you in the ticket thread, and I will be glad to help you amending any Italian captions once this version is released.

activescott commented:
This is good for me now. Here is how you can test this for future reference:

CI always publishes artifacts for every build. So for the most recent build on this PR at the time I'm writing this comment they are at https://ci.appveyor.com/project/activescott/lessmsi/builds/50698795/artifacts

- You can download the zip file there and check inside of it and see the file `\it\lessmsi-gui.resources.dll`.
- You can download the nuspec file there (or build it locally) and then install it with choco locally with the command`choco install lessmsi --version="0.0.0" --source="C:\src\lessmsi\src\.deploy\chocolateypackage"` (in this case I used version 0.0.0 build locally). Then it will print out where it installs it and in that directory I can see the it subdirectory with the expected resources file there.

activescott commented:
@mega5800 Please confirm all tasks are complete here and that you want to merge it and I'll merge it.

activescott commented:
Merged. Sorry for the delay. I was/am traveling for work.

activescott commented:
:tada: This PR is included in version 2.2.0 :tada:

The release is available on:
- [GitHub release](https://github.com/activescott/lessmsi/releases/tag/v2.2.0)
- `v2.2.0`

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
